### ISSUE TITLE
Query UserTask for user

### ISSUE THREAD

**Issue by IntranetFactory** [elsa-workflows/elsa-core#241]
We have a document approval workflow with e.g. document1 to be approved by userA, document2 to be approved by userB and document3 to be approved by userA    
    
I'd like to get a list of pending approvals for userA, but didn't find a way to query BlockingActivityEntity for a given user.    
    
Would it make sense to add a string BlockedById (or similar) property to BlockingActivityEntity? This would allow to store an id to identify what external resource is currently blocking the task.

**Comment by sfmskywalker** [comment]
I think adding e.g. `BlockedById` to `BlockingActivityEntity` might be a little bit too arbitrary from the perspective of a blocking activity, but your use case is a very valid one and something I still need to design for.

Adding some `UserId` field to the `UserTask` activity itself might make sense though. Then you would be able to at the very least list all workflow instances blocked by the `UserTask` activity like so:

```csharp
var tuples = await workflowInstanceStore.ListByBlockingActivityAsync("UserTask");
```

The next step would be to filter the result by `userA` as follows:

```csharp
var pendingApprovals = 
   from (workflowInstance, activityInstance) in pendingApprovals
   let userId = activityInstance.State["UserId"]
   where userId == "userA"
   select workflowInstance
```

A downside with this is might be the fact that you can't do the full filtering on the database server itself; you need to fetch all workflow instances blocked by "UserTask" first, which might include way more records than the ones you're interested in, so you need to filter the rest in memory as shown above.

To mitigate this, we might consider also storing activity state in a normalized fashion (one row containing `| workflow instance ID | activity ID | activity type | activity state key | activity state value`)

Then you could perform a DB query like:

```csharp
var pendingApprovals = workflowInstanceStore.QueryByActivityStateAsync(activityType: "UserTask", key: "UserId": value: "userA");
```

Yet another approach could be that your application listens for some "Workflow Suspended" event, storing records in a custom table called "PendingApprovals" containing the workflow instance ID and user ID. You would then also need to handle some "Workflow Resumed" event to remove from "PendingApprovals".

Thoughts?

**Comment by IntranetFactory** [comment]
After I saw the CorrelationId on the instance level, I thought that having a generic name might be the preferred naming convention.    

We plan to add additional blocking types - so like GroupTask (where the task will be assigned to a group of users) or HelpdeskTask (where an ticket in an helpdesk system is created/selected) and the workflow needs to wait until the ticket is closed. 

So I was even thinking about  the need to have two properties `BlockedById` and `BlockedByType`.

**Comment by sfmskywalker** [comment]
You make a good point. `CorrelationId` is intentionally generic, but intended to be able to associate a workflow instance with the identifier of an application-specific object (such as a document). Perhaps having a `Tag` property on `BlockingActivity` isn't such a bad idea, as it would probably cover most use cases similar to that of yours. The activities suspending the workflow would return a `SuspendResult` that carries whichever tag value that's appropriate for a given activity.

For example, UserTask would do: `return Suspend(UserId)`, and your custom `GroupTask` would do: `return Suspend(GroupName)` (UserId & GroupName would both be of type `IWorkflowExpression<string>`.

Then `IWorkflowInstanceStore` would allow you to make very specific queries, such as `workflowInstanceStore.ListByBlockingActivityAsync(activityType: "GroupTask", tag: "Moderators")`.

What do you think?

**Comment by IntranetFactory** [comment]
That would be a perfect solution for all of our use cases.

**Comment by sniperwolfpk5** [comment]
@IntranetFactory @sfmskywalker   how to trigger user task. There is no sample I tried to trigger using workflows/trigger/{token} but it didn't work. Can you give me short sample. 

Please note I want to achieve it by using dashboard not by code.

Regards,

**Comment by sfmskywalker** [comment]
@sniperwolfpk5 Please open a separate issue with your question. I think it's a good one, and want to keep track of it and provide a complete guide for it in the documentation. Thanks!
